File: 2week/지연/4.next-permutation.py
from typing import List

# 순열 찾기


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        end_idx = max_idx = len(nums) - 1

        # 큰 수 위치 찾기
        while max_idx > 0 and nums[max_idx - 1] >= nums[max_idx]:
            max_idx -= 1

        # nums are in descending order
        if max_idx == 0:
            nums.reverse()
            return nums

        # find the last "ascending" position
        swap_idx = max_idx - 1
        while nums[end_idx] <= nums[swap_idx]:
            end_idx -= 1
        nums[swap_idx], nums[end_idx] = nums[end_idx], nums[swap_idx]

        # reverse the second part
        max, min = swap_idx + 1, len(nums) - 1
        while max < min:
            nums[max], nums[min] = nums[min], nums[max]
            max += 1
            min -= 1
        return nums


print(Solution().nextPermutation([1, 1, 2, 6, 5]))
print(Solution().nextPermutation([3, 2, 1]))
print(Solution().nextPermutation([1, 2, 8, 6, 5]))

File: 2.brute-force/동규/4.next-permutation.js
function nextPermutation(nums) {
  const swap = (arr, a, b) => ([arr[a], arr[b]] = [arr[b], arr[a]]);

  let i = nums.length - 2; //1
  while (i >= 0 && nums[i] >= nums[i + 1]) i--;

  if (i >= 0) {
    let j = nums.length - 1;
    while (nums[i] >= nums[j]) j--;
    swap(nums, i, j);
  }
  nums.push(...nums.splice(i + 1).reverse());
}

nextPermutation([1, 2, 3]);
